VarExpand for the register runtime

systay opened the pull request.

Commit:
Add var-length expand to the register runtime

Commit:
Make register varexpand produce lists of relationships

Commit:
Stop varexpand from being used when predicates are available

systay commented:
This is built on top of #9759 to show which scenarios are successful after this change

Commit:
Mark tests as solved

craigtaverner commented on enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala:
```diff
@@ -41,6 +41,21 @@ class MatchAcceptanceTest extends ExecutionEngineFunSuite with QueryStatisticsTe
     result.toList should equal(List(Map("count(n.name)" -> 3)))
   }
 
+  test("varlength apa") {
```
Seems like a non-ideal test name....

craigtaverner replied:
Perhaps this was a development-only test? Not meant to be committed?

systay replied:
Yep

craigtaverner commented on enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/RegisterAllocation.scala:
```diff
@@ -141,6 +141,12 @@ object RegisterAllocation {
         cartesianProductPipeline
 
+      case VarExpand(lhs, IdName(from), dir, projectedDir, types, IdName(to), IdName(edge), length, ExpandAll, predicates) =>
+        val pipeline = allocate(lhs, nullable, argument)
```
Why doesn't this do the same as expand and create a new pipe using deepClone? Since the VarExpand, like the normal Expand, will change the number of columns, we would expect a new pipeline, right?

systay replied:
Absolutely right

craigtaverner commented on enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/interpreted/RegisteredPipeBuilder.scala:
```diff
@@ -94,6 +95,22 @@ class RegisteredPipeBuilder(fallback: PipeBuilder,
+      case VarExpand(_, IdName(fromName), dir, projectedDir, types, IdName(toName), IdName(relName), VarPatternLength(min, max), expansionMode, predicates) =>
+        // TODO: This is not right!
+        if(predicates.nonEmpty)
+          throw new CantCompileQueryException("does not handle varexpand with predicates")
```
Other PR's handing OptionalExpandRegisterPipe have solutions for predicates, so we should reuse those.

systay replied:
Very different. I have been working on solving this, but it can be done as a separate step. I'll push what I have so you can see what I'm thinking about.

craigtaverner commented on enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/interpreted/pipes/VarLengthExpandRegisterPipe.scala:
```diff
@@ -0,0 +1,135 @@
+            val resultRow = PrimitiveExecutionContext(pipeline)
+            resultRow.copyFrom(inputRowWithFromNode)
+            resultRow.setLongAt(toOffset, toNode)
+            val rels = relIds.map(state.query.relationshipOps.getById)
```
Not great to have to hit the store to re-lookup the relationships again, when we know the previous lookup already got all this information.

systay replied:
+1

craigtaverner commented on enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/interpreted/pipes/VarLengthExpandRegisterPipe.scala:
```diff
@@ -0,0 +1,135 @@
+  }
+
+  private def isToNodeValid(row: ExecutionContext, node: LNode): Boolean =
+    !closedPath || row.getLongAt(toOffset) == node
```
The boolean closedPath is non-intuitive. Would be better with something like 'endNodeBound', or even 'ExpandInto'

craigtaverner commented on enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/LernaeanTestSupport.scala:
```diff
@@ -87,7 +87,7 @@ trait LernaeanTestSupport extends CypherTestSupport {
 
   private def extractFirstScenario(config: TestConfiguration): TestScenario = {
-    val preferredScenario = Scenarios.CommunityInterpreted
+    val preferredScenario = Scenarios.CommunityInterpreted3_3
```
This seems to be unrelated to this work, and also seems incomplete, because the same improvement is not done to EnterpriseInterpreted or Compiled.

craigtaverner reviewed (changes requested):
I did not find any tests for this work. Presumably at least the RegisterPipeBuilderTest should be enhanced to cover this?

craigtaverner committed:
Removed development-only test

craigtaverner committed:
New test for Registered VarExpand, and fixed pipeline allocation

craigtaverner committed:
Improved allocation testing for varlength expand

craigtaverner committed:
Clearer parameter naming for expand-all in var expand

craigtaverner committed:
Get register ast expressions to print themselves correctly in plan descriptions

craigtaverner committed:
Reduced kernel interaction for relationship lookup in var-expand

Commit:
Compilation errors

Commit:
Test failures

craigtaverner committed:
Reduced object creation and kernel interaction for relationships

VarExpand was re-creating the same relationship obejcts many times, as well as creating them with only rel-id
so that any later interaction would lilely result in kernel calls. This way we keep all other ids, so none need
to be looked up.

craigtaverner committed:
Revert some text code to original
